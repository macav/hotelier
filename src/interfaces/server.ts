export enum Status {
  RUNNING = 'running',
  RESTARTING = 'restarting',
  STOPPED = 'stopped',
  CRASHED = 'crashed',
}
export interface Server {
  id: string;
  status: Status;
  env: { [key: string]: any };
}
