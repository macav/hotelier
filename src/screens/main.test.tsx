import React from 'react';
import renderer from 'react-test-renderer';
import Main from './main';

import mockFetch from 'jest-fetch-mock';
import HotelApi from '../api';
import { Status } from '../interfaces';

jest.mock('react-transition-group');

describe('Main', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  const mockServers = {
    server1: { status: 'running' },
    server2: { status: 'stopped' },
  };

  function mockSuccessfulResponse() {
    mockFetch.mockResponse(JSON.stringify(mockServers));
  }

  it('matches the snapshot', (done) => {
    mockSuccessfulResponse();
    const rendered = renderer.create(<Main />);
    expect(rendered.toJSON()).toMatchSnapshot('loading');
    process.nextTick(() => {
      expect(rendered.toJSON()).toMatchSnapshot('loaded');
      done();
    });
  });

  it('updates the state with the servers loaded from the API', (done) => {
    mockSuccessfulResponse();
    const instance = (renderer.create(<Main />).getInstance() as any) as Main;
    expect(instance.state.loading).toBe(true);
    process.nextTick(() => {
      expect(instance.state.servers.length).toEqual(2);
      expect(instance.state.servers[0].id).toEqual('server1');
      expect(instance.state.servers[0].status).toEqual('running');
      expect(instance.state.servers[1].id).toEqual('server2');
      expect(instance.state.servers[1].status).toEqual('stopped');
      expect(instance.state.loading).toBe(false);
      done();
    });
  });

  it('adds a new server to the state', () => {
    mockSuccessfulResponse();
    const instance = (renderer.create(<Main />).getInstance() as any) as Main;
    mockFetch.mockResponse(JSON.stringify({ server3: { status: 'running' } }));
    jest.runTimersToTime(3000);
    process.nextTick(() => {
      expect(instance.state.servers.length).toEqual(1);
      expect(instance.state.servers[0].id).toEqual('server3');
      expect(instance.state.servers[0].status).toEqual('running');
      expect(instance.state.loading).toBe(false);
    });
  });

  it('reloads servers on event', () => {
    mockSuccessfulResponse();
    (global as any).ipcRenderer = {
      on: (name: string, callback: any) =>
        callback(null, { ...mockServers, server3: { status: 'running' } }),
    };
    const instance = (renderer.create(<Main />).getInstance() as any) as Main;
    expect(instance.state.servers.length).toEqual(3);
    expect(instance.state.servers).toEqual(
      expect.arrayContaining([expect.objectContaining({ id: 'server3' })])
    );
  });

  it('renders empty list when we get error from the api', (done) => {
    console.error = jest.fn();
    try {
      mockFetch.mockResponse('', { status: 500 });
      const instance = (renderer.create(<Main />).getInstance() as any) as Main;
      process.nextTick(() => {
        expect(instance.state.servers).toEqual([]);
        expect(console.error).toHaveBeenCalled();
        done();
      });
    } catch (e) {
      done();
    }
  });

  describe('#updateServerStatus', () => {
    beforeEach(() => {
      mockSuccessfulResponse();
    });

    it('updates the corresponding server with the new status', (done) => {
      const instance = (renderer.create(<Main />).getInstance() as any) as Main;
      process.nextTick(() => {
        instance.updateServerStatus('server1', Status.STOPPED);
        expect(instance.state.servers[0].status).toEqual(Status.STOPPED);
        done();
      });
    });
  });
});
