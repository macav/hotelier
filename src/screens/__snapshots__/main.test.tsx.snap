// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main matches the snapshot: loaded 1`] = `
<div
  className="main-container"
>
  <div
    className="header-arrow"
  />
  <div
    className="window"
  >
    <nav
      className="navbar navbar-expand-lg navbar-dark bg-primary"
    >
      <div
        className="mx-auto font-weight-bold text-center text-white"
      >
        Hotelier
      </div>
    </nav>
    <div
      className="window-content"
    >
      <div
        className="pane"
      >
        <div>
          <ul
            className="list-group"
          >
            <li
              className="list-group-item list-group-item-action server-item"
            >
              <div
                className="d-flex justify-content-between"
              >
                <div
                  className="server-name server-name--running"
                  onClick={[Function]}
                >
                  server1
                </div>
                <div
                  className="float-right"
                >
                  <div
                    className="d-inline"
                  />
                  <div
                    className="d-inline mx-1"
                  >
                    <button
                      className="btn toggle-state-button btn-danger text-white"
                      onClick={[Function]}
                      title="Toggle"
                      type="button"
                    >
                      <span
                        className="tcon tcon-remove tcon-remove--chevron-right "
                      />
                    </button>
                  </div>
                  <div
                    className="d-inline"
                  >
                    <button
                      className="btn btn-light mr-1"
                      onClick={[Function]}
                      title="Logs"
                      type="button"
                    >
                      <i
                        className="fas fa-bars"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </li>
            <li
              className="list-group-item list-group-item-action server-item"
            >
              <div
                className="d-flex justify-content-between"
              >
                <div
                  className="server-name server-name--stopped"
                  onClick={[Function]}
                >
                  server2
                </div>
                <div
                  className="float-right"
                >
                  <div
                    className="d-inline"
                  />
                  <div
                    className="d-inline mx-1"
                  >
                    <button
                      className="btn toggle-state-button btn-light"
                      onClick={[Function]}
                      title="Toggle"
                      type="button"
                    >
                      <span
                        className="tcon tcon-remove tcon-remove--chevron-right tcon-transform"
                      />
                    </button>
                  </div>
                  <div
                    className="d-inline"
                  >
                    <button
                      className="btn btn-light mr-1"
                      onClick={[Function]}
                      title="Logs"
                      type="button"
                    >
                      <i
                        className="fas fa-bars"
                      />
                    </button>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
    <footer
      className="fixed-bottom bg-primary"
    >
      <div
        className="float-left"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          Hotel
        </button>
      </div>
      <div
        className="float-right"
      >
        <div
          className="btn-group"
        >
          <button
            className="btn btn-primary"
            id="close-app"
            onClick={[Function]}
          >
            <span
              className="fas fa-power-off"
              title="Quit"
            />
          </button>
        </div>
      </div>
    </footer>
  </div>
</div>
`;

exports[`Main matches the snapshot: loading 1`] = `
<div
  className="main-container"
>
  <div
    className="header-arrow"
  />
  <div
    className="window"
  >
    <nav
      className="navbar navbar-expand-lg navbar-dark bg-primary"
    >
      <div
        className="mx-auto font-weight-bold text-center text-white"
      >
        Hotelier
      </div>
    </nav>
    <div
      className="window-content"
    >
      <div
        className="pane"
      >
        <div
          className="summary"
        >
          Loadingâ€¦
        </div>
        <div>
          <ul
            className="list-group"
          />
        </div>
      </div>
    </div>
    <footer
      className="fixed-bottom bg-primary"
    >
      <div
        className="float-left"
      >
        <button
          className="btn btn-primary"
          onClick={[Function]}
        >
          Hotel
        </button>
      </div>
      <div
        className="float-right"
      >
        <div
          className="btn-group"
        >
          <button
            className="btn btn-primary"
            id="close-app"
            onClick={[Function]}
          >
            <span
              className="fas fa-power-off"
              title="Quit"
            />
          </button>
        </div>
      </div>
    </footer>
  </div>
</div>
`;
